src/sycltest/plugin-Test1/gpuAlgo1.cc

```diff
@@ -0,0 +1,175 @@
+    int row = item_ct1.get_group(1) * item_ct1.get_local_range().get(1) + item_ct1.get_local_id(1);
+    int col = item_ct1.get_group(2) * item_ct1.get_local_range().get(2) + item_ct1.get_local_id(2);
+
+    if (row < numElements && col < numElements) {
+      T tmp = 0;
+      for (int i = 0; i < numElements; ++i) {
+        tmp += a[row * numElements + i] * b[i * numElements + col];
+      }
+      c[row * numElements + col] = tmp;
+    }
+  }
+
+  template <typename T>
+  SYCL_EXTERNAL void matrixMulVector(const T *a, const T *b, T *c, int numElements, sycl::nd_item<3> item_ct1) {
+    int row = item_ct1.get_group(1) * item_ct1.get_local_range().get(1) + item_ct1.get_local_id(1);
+
+    if (row < numElements) {
+      T tmp = 0;
+      for (int i = 0; i < numElements; ++i) {
+        tmp += a[row * numElements + i] * b[i];
+      }
+      c[row] = tmp;
+    }
+  }
+}  // namespace
+
+using namespace gpu_algo_1;
+
+constexpr unsigned int sqrt2(unsigned int value) {
+  unsigned int result = 1;
+  while (value >= 4) {
+    value /= 4;
+    result *= 2;
+  }
+  return result;
+}
+
+cms::sycltools::device::unique_ptr<float[]> gpuAlgo1(sycl::queue stream) {
+  // query the device for the maximum "block size" or workgroup size
+  const unsigned int workgroupSize = stream.get_device().get_info<sycl::info::device::max_work_group_size>();
+
+  // build a rectangular "block" or workgroup
+  const unsigned int workgroupRectB = sqrt2(workgroupSize);
+  const unsigned int workgroupRectA = workgroupSize / workgroupRectB;
+
+  auto h_a = cms::sycltools::make_host_unique<float[]>(NUM_VALUES, stream);
+  auto h_b = cms::sycltools::make_host_unique<float[]>(NUM_VALUES, stream);
+
+  for (auto i = 0; i < NUM_VALUES; i++) {
+    h_a[i] = i;
+    h_b[i] = i * i;
+  }
+
+  auto d_a = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+  auto d_b = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+
+  stream.memcpy(d_a.get(), h_a.get(), NUM_VALUES * sizeof(float));
+  stream.memcpy(d_b.get(), h_b.get(), NUM_VALUES * sizeof(float));
+
+  unsigned int threadsPerBlock = std::min<unsigned int>(NUM_VALUES, workgroupSize);
+  unsigned int blocksPerGrid = (NUM_VALUES + threadsPerBlock - 1) / threadsPerBlock;
+
+  std::cerr << "block size: " << threadsPerBlock << " / " << stream.get_device().get_info<sycl::info::device::max_work_group_size>() << std::endl; 
+  std::cerr << "grid size:  " << blocksPerGrid   << std::endl;
+
+  auto d_c = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+  stream.submit([&](sycl::handler &cgh) {
+    auto d_a_get_ct0 = d_a.get();
+    auto d_b_get_ct1 = d_b.get();
+    auto d_c_get_ct2 = d_c.get();
+
+    cgh.parallel_for(sycl::nd_range<3>(sycl::range<3>(1, 1, blocksPerGrid) * sycl::range<3>(1, 1, threadsPerBlock),
+                                       sycl::range<3>(1, 1, threadsPerBlock)),
+                     [=](sycl::nd_item<3> item_ct1) {
```

Comment:
Could one do
```suggestion
    cgh.parallel_for(sycl::nd_range<1>(sycl::range<1>(blocksPerGrid) * sycl::range<1>(threadsPerBlock),
                                       sycl::range<3>(threadsPerBlock)),
                     [=](sycl::nd_item<1> item_ct1) {
```
?

Reply:
Yes, but it should be simplified a lot more than that :-)